File: worker/src/controllers/stockController.ts
import { Request, Response } from "express";
import { STOCK_BALANCES } from "../models/balances";
import { ORDERBOOK } from "../models/orderbook";

export const createSymbol = (stockSymbol: string) => {
  // const defaultStockBalance: StockBalance = {
  //     yes: {
  //         quantity: 0,
  //         locked: 0
  //     },
  //     no: {
  //         quantity: 0,
  //         locked: 0
  //     }
  // };

  // Object.keys(STOCK_BALANCES).forEach(userId => {
  //     STOCK_BALANCES[userId][stockSymbol] = {
  //         ...defaultStockBalance,
  //         yes: {
  //             ...defaultStockBalance.yes,
  //             ...{ quantity: defaultStockBalance.yes!.quantity, locked: defaultStockBalance.yes!.locked }
  //         },
  //         no: {
  //             ...defaultStockBalance.no,
  //             ...{ quantity: defaultStockBalance.no!.quantity, locked: defaultStockBalance.no!.locked }
  //         }
  //     };
  // });

  Object.keys(STOCK_BALANCES).forEach((userId) => {
    STOCK_BALANCES[userId][stockSymbol] = {
      yes: {
        quantity: 0,
        locked: 0,
      },
      no: {
        quantity: 0,
        locked: 0,
      },
    };
  });

  ORDERBOOK[stockSymbol] = { yes: {}, no: {} };

  return { error: false, msg: `Symbol ${stockSymbol} created successfully.` };
};

export const getStockBalance = () => {
  return STOCK_BALANCES;
};

export const getIndividualStockBalance = (userId: string) => {
  const stock = STOCK_BALANCES[userId];
  if (!stock) {
    return { error:true, msg: "User not found or no stock balance available" };
  }
  return {error:false, msg: stock};
};

export const mintTrade = (userId: string, stockSymbol: string, quantity: number) => {
  if (!STOCK_BALANCES[userId]) {
    STOCK_BALANCES[userId] = {};
  }

  if (!STOCK_BALANCES[userId][stockSymbol]) {
    STOCK_BALANCES[userId][stockSymbol] = {
      yes: {
        quantity: 0,
        locked: 0,
      },
      no: {
        quantity: 0,
        locked: 0,
      },
    };
  }

  if (!STOCK_BALANCES[userId][stockSymbol].yes || !STOCK_BALANCES[userId][stockSymbol].no) {
    STOCK_BALANCES[userId][stockSymbol].yes = { quantity: 0, locked: 0 };
    STOCK_BALANCES[userId][stockSymbol].no = { quantity: 0, locked: 0 };
  }

  STOCK_BALANCES[userId][stockSymbol].yes.quantity += quantity;
  STOCK_BALANCES[userId][stockSymbol].no.quantity += quantity;

  return { error: false, msg: STOCK_BALANCES[userId][stockSymbol] };
};

File: go_worker/controllers/stockController.go
package controllers

import (
	"fmt"

	"github.com/dipenbhat557/probo_clone/models"
)

func CreateSymbol(stockSymbol string) map[string]interface{} {
	for userId := range models.STOCK_BALANCES {
		models.STOCK_BALANCES[userId][stockSymbol] = models.UserStock{
			Yes: models.Stock{
				Quantity: 0,
				Locked:   0,
			},
			No: models.Stock{
				Quantity: 0,
				Locked:   0,
			},
		}
	}

	models.ORDERBOOK[stockSymbol] = models.OrderBookSymbol{
		Yes: make(map[float64]models.PriceLevel),
		No:  make(map[float64]models.PriceLevel),
	}

	return map[string]interface{}{
		"error": false,
		"msg":   fmt.Sprintf("Symbol %s created successfully.", stockSymbol),
	}
}

func GetStockBalance() map[string]interface{} {
	return map[string]interface{}{"error": false, "msg": models.STOCK_BALANCES}
}

func GetIndividualStockBalance(userId string) map[string]interface{} {
	stock, exists := models.STOCK_BALANCES[userId]
	if !exists {
		return map[string]interface{}{
			"error": true,
			"msg":   "User not found or no stock balance available",
		}
	}
	return map[string]interface{}{
		"error": false,
		"msg":   stock,
	}
}

func MintTrade(userId string, stockSymbol string, quantity int) map[string]interface{} {

	if _, exists := models.STOCK_BALANCES[userId]; !exists {
		models.STOCK_BALANCES[userId] = make(map[string]models.UserStock)
	}

	if _, exists := models.STOCK_BALANCES[userId][stockSymbol]; !exists {
		models.STOCK_BALANCES[userId][stockSymbol] = models.UserStock{
			Yes: models.Stock{
				Quantity: 0,
				Locked:   0,
			},
			No: models.Stock{
				Quantity: 0,
				Locked:   0,
			},
		}
	}

	indStock := models.STOCK_BALANCES[userId][stockSymbol]
	indStock.Yes.Quantity += quantity
	indStock.No.Quantity += quantity
	models.STOCK_BALANCES[userId][stockSymbol] = indStock

	return map[string]interface{}{
		"error": false,
		"msg":   models.STOCK_BALANCES[userId][stockSymbol],
	}
}
